compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -365,46 +366,94 @@ impl<'a> AstValidator<'a> {
         }
     }
 
-    /// An `extern "custom"` function must be unsafe, and must not have any parameters or return
-    /// type.
-    fn check_custom_abi(&self, ctxt: FnCtxt, ident: &Ident, sig: &FnSig) {
+    /// Check that this function does not violate the constraints of its ABI.
+    fn check_abi(&self, abi: ExternAbi, ctxt: FnCtxt, ident: &Ident, sig: &FnSig) {
+        match AbiMap::from_target(&self.sess.target).canonize_abi(abi, false) {
+            AbiMapping::Direct(canon_abi) | AbiMapping::Deprecated(canon_abi) => {
+                match canon_abi {
+                    CanonAbi::C
+                    | CanonAbi::Rust
+                    | CanonAbi::RustCold
+                    | CanonAbi::Arm(_)
+                    | CanonAbi::GpuKernel
+                    | CanonAbi::X86(_) => { /* nothing to check */ }
+
+                    CanonAbi::Custom => {
+                        // An `extern "custom"` function must be unsafe.
+                        self.check_abi_is_unsafe(abi, ctxt, sig);
+
+                        // An `extern "custom"` function cannot be `async` and/or `gen`.
+                        self.check_abi_is_not_coroutine(abi, sig);
+
+                        // An `extern "custom"` function must have type `fn()`.
```

Comment:
pedantic note: its type is `unsafe extern "custom" fn()`. and yes, you can have *different* implementations on `unsafe fn()` and `fn()`.

don't actually change anything, I just couldn't resist noting.